### ISSUE TITLE
Unit tests

### ISSUE THREAD

**Issue by larskotthoff** [LinlinYin/mlr3vis#2]
Please add some tests with testthat. You can copy some of the tests from the mlr repo to get started.

**Comment by LinlinYin** [comment]
Thanks! I will do it.

**Comment by LinlinYin** [comment]
After reading the documents for testthat package, I didn't find anything related to graphic output. May I know what is the best way to test graphic output? A function like:   expect_equal(ggplot(...), expectedFigure)? Thanks!

**Comment by pat-s** [comment]
You can take look into the _ggplot2_ package how it is handled there.

**Comment by larskotthoff** [comment]
We test the plots by running the functions, saving the output as SVG, and parsing it to check that elements that should be there are present (for example the name of the learner, or the expected number of bars). See https://github.com/mlr-org/mlr/blob/master/tests/testthat/test_base_plotLearnerPrediction.R for an example.

**Comment by pat-s** [comment]
I am not sure if this is the right way to do it, we should really take a look how ggplot2 is doing it themselves? They should know best. Right now would be a good time to do this.

**Comment by larskotthoff** [comment]
Let's have a look at both and see what they can and can't do and then decide.

**Comment by LinlinYin** [comment]
My understanding for this: Do we really want the exactly same figure as expected? Is it better/easier if we only confirm that there is a figure object generated, which means our code is running (maybe not running perfectly, but it is enough for testing purpose?).

**Comment by larskotthoff** [comment]
We need to go a bit beyond that. In particular, what we're currently checking in our tests is whether the objects we expect to be there are in fact there (e.g. 10 points if we're plotting 10 predictions). Just verifying that the plot object is created isn't enough because the plot might be wrong and we would never even notice.

**Comment by pat-s** [comment]
Lars is right here, the plot creation is the first step only.
Sometimes changes in the code cause certain plots to miss the actual observations, the result is misaligned or the headings are wrong.

In an optimal way, every part of the resulting plot should be checked so that we catch possible leftovers when making changes of algorithm names that are passed to the final plot and depend on _mlr3_ internal naming (just an example).

For now we can start with the first layer to get this process started in the first place.

**Comment by larskotthoff** [comment]
@LinlinYin please have a look at this and come up with a few alternatives for testing as soon as possible. We need to make progress on this.

**Comment by LinlinYin** [comment]
I read the examples in ggplot2 and think they had a good idea about it. Please see example below from ggplot testing boxplot, from [ggplot2 github](https://github.com/tidyverse/ggplot2/blob/master/tests/testthat/test-geom-boxplot.R)

> test_that("geom_boxplot range includes all outliers", {
>   dat <- data_frame(x = 1, y = c(-(1:20) ^ 3, (1:20) ^ 3) )
>   p <- ggplot_build(ggplot(dat, aes(x,y)) + geom_boxplot())
> 
>   miny <- p$layout$panel_params[[1]]$y.range[1]
>   maxy <- p$layout$panel_params[[1]]$y.range[2]
> 
>   expect_true(miny <= min(dat$y))
>   expect_true(maxy >= max(dat$y))
> })

**Comment by LinlinYin** [comment]
Pleas review the test_plotBMRBoxplots.R in test folder, based on the ggplot2 test code listed above. Thanks!~

**Comment by larskotthoff** [comment]
Ok, this approach looks good. Is there a way of verifying how many elements are being plotted?

**Comment by LinlinYin** [comment]
I am not sure if I understand the "element" correctly. I think it is number of boxplot in the figure? Also got some idea about it after reading ggplot2 example. 
Please review the changes in `test_plotBMRBoxplots.R`.

Some key functions to test the number of box in the figure:
>   p = plotBMRBoxplots(bmr)
>   pLayerData <- layer_data(p)
> 
>   expect_true(nrow(pLayerData) ==6)

**Comment by larskotthoff** [comment]
Yes, it sounds like this would be the right test. Can you please be specific as to which change you are referring to?

**Comment by LinlinYin** [comment]
See code in `test_plotBMRBoxplots.R` below:

> test_that("plotBMRBoxplots one example contains six boxplot", {
>   tasks = mlr_tasks$mget(c("pima", "sonar", "spam"))
>   learners = mlr_learners$mget(c("classif.featureless", "classif.rpart"))
>   resamplings = mlr_resamplings$mget("cv")
>   design = expand_grid(tasks, learners, resamplings)
>   bmr = benchmark(design)
> 
>   p = plotBMRBoxplots(bmr)
>   pLayerData <- layer_data(p)
> 
>   expect_true(nrow(pLayerData) == 6)
> })

**Comment by larskotthoff** [comment]
Looks good, but please replace the hardcoded 6 with the product of the number of learners and tasks.

And PLEASE follow the style guide...

**Comment by LinlinYin** [comment]
Thanks! It has been changed into `length(tasks) * length(learners)`
Yes, the mlr_style was also applied in this file.
